File: solidPrinciples/(L)Fixed.ts
abstract class Bird {
  abstract info(): void;
}

class FlyingBird extends Bird {
  fly(): void {
    console.log("Flying...");
  }
}

class Sparrow extends FlyingBird {
  info(): void {
    console.log("I am a sparrow.");
  }
}

class Penguin extends Bird {
  info(): void {
    console.log("I am a penguin. I can't fly.");
  }
}

const sparrow = new Sparrow();
sparrow.fly(); // Output: Flying...
sparrow.info(); // Output: I am a sparrow.

const penguin = new Penguin();
penguin.info(); // Output: I am a penguin. I can't fly.

File: solidPrinciplesCpp/(L)Fixed.cpp
#include <iostream>

class Bird
{
public:
    virtual void info()
    {
        std::cout << "I am a bird.\n";
    }
};

class FlyingBird : public Bird
{
public:
    virtual void fly()
    {
        std::cout << "Flying...\n";
    }
};

class Sparrow : public FlyingBird
{
};

class Penguin : public Bird
{
    void info() override
    {
        std::cout << "I am a penguin. I can't fly.\n";
    }
};

int main()
{
    Sparrow sparrow;
    Penguin penguin;

    sparrow.fly(); // Output: "Flying..."

    penguin.info(); // Output: "I am a penguin. I can't fly."
}
